synapse/handlers/admin.py

```diff
@@ -29,10 +32,106 @@
 class AdminHandler(BaseHandler):
     def __init__(self, hs):
         super(AdminHandler, self).__init__(hs)
-
         self.storage = hs.get_storage()
         self.state_store = self.storage.state
 
+    async def validate_admin_token(
+        self, servlet, request, raise_if_missing: bool = True
+    ) -> bool:
+        """
+        Validate that there is an admin token on the request, and that it can
+        access this servlet.
+        """
+        # This servlet can't be validated by an admin token. Error out.
+        if servlet.PERMISSION_CODE is None:
```

Comment:
probably clearer to pass in the desired permission code rather than a reference to the entire servlet?

Reply:
the idea is that a REST-ful API will generally have the type of things (such as users) as the endpoints, and then you can declare that you can PUT or DELETE or GET a user, or whatever, so that's what this is trying to aim for by looking on the servlet

Reply:
I think it leads to an unnecessarily tight coupling between the auth code and the servlet impl.